### ISSUE TITLE
PLINK Error: File read failure for all markers using local VCFs

### ISSUE THREAD

**Issue by oyhel** [welchr/swiss#3]
#### Command-line used to cause the issue
```
        swiss \
            --assoc {input} \
            --pval-col PVALUE \
            --ld-clump \
            --ld-clump-source vcfmap.json \
            --ld-gwas-source 1000G_2014-11_EUR \
            --clump-ld-thresh 0.5 \
            --clump-p 5e-06 \
            --gwas-cat ebi \
            --build hg19 \
            --include-cols "STUDYBETA" \
            --out {output}
```

#### Path to your data (if publicly available)

The data is not publicly available but here is a sample of my VCF (first 5 columns):
#CHROM  POS     ID      REF     ALT
1       762320  rs75333668      C       T
1       762409  rs563839199     G       C
1       762856  rs142969342     G       A
1       763394  rs3115847       G       A

And sample of the .assoc file
CHR     POS     REF     ALT     MARKER_ID       PVALUE  STUDYBETA
1      762409  G       C       rs563839199     0.334595        0.0727147
1      763394  G       A       rs3115847       0.752615        0.00531069
1      766007  A       C       rs61768174      0.74356 0.0079206
1      768253  A       C       rs2977608       0.992767        0.0001617
1      768448  G       A       rs12562034      0.491016        0.0181229
1      768819  C       T       rs12562811      0.749079        -0.0307074

#### Problem description

When I try to clump using my own VCFs, all markers fail to clump. When I use eg. 1000G_2014-11_EUR most pass, but some fail. This is understandable since all markers are not in the reference. Looking at the PLINK error it says Error: File read failure for all markers. By looking into the code in the PlinkLDFinder class I suspect the problem might be that swiss is asking PLINK to extract the marker by EPACTS-id. But my VCFs only have rsIDs which I suspect is why it fails for all markers. 

#### Swiss version
1.04b

**Comment by welchr** [comment]
Hmm, it should ignore the ID column and simply use CHR/POS/REF/ALT when matching variant names. Can you post the error message + stack trace if there's one?

**Comment by oyhel** [comment]
Unfortunately there is not stack trace since this happens in PLINK and python does not seem to break. With the command above swiss just blasts over all the markers after filtering on the highest P-values (I assume). Then error message for all the markers (and no overall statistics on sample count etc. so it seems it just never manages to create the file?):

```
PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/
(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /media/local-disk2/helg/swiss-test/tmpcYyGeq.log.
Options in effect:
  --ld-snp id0
  --ld-window 99999
  --ld-window-kb 99999
  --ld-window-r2 0.8
  --out /media/local-disk2/helg/swiss-test/tmpcYyGeq
  --r2 gz dprime with-freqs yes-really
  --threads 1
  --vcf /dev/fd/0

128829 MB RAM detected; reserving 64414 MB for main workspace.

Error: File read failure.

Warning: could not calculate LD for variant 2:25191745_A/T at 2:25191745
```
It then returns a long list with fail for all markers tested (list below only shows the first lines).

```
CHR     POS     REF     ALT     MARKER_ID       PVALUE  STUDYBETA       variant TRAIT   ld_with ld_with_values  failed_clump
1       65893960        A       G       rs4468199       2.16985e-06     0.128909        1:65893960_A/G  NA                      fail
1       65894342        C       T       rs3790433       3.3609e-08      0.0975544       1:65894342_C/T  NA                      fail
1       65894940        A       G       rs75465244      4.32741e-06     0.124989        1:65894940_A/G  NA                      fail
1       65895130        A       G       rs7534511       4.8873e-06      0.0755236       1:65895130_A/G  NA                      fail
1       65895927        T       C       rs9436301       2.85282e-09     0.108501        1:65895927_T/C  NA                      fail
```

If you have any suggestions I could try, please let me know! Great tool once I get it working! ;)

**Comment by welchr** [comment]
Can you try running swiss with `--debug` - do you see a stack trace? If so can you include it? The frustrating part is I know I've run into this issue before, but I can't recall the solution. Hopefully the debug output will shed some light.

**Comment by oyhel** [comment]
This is the stack trace when running with --debug

```
Working on LD clumping variant 1:65991203_A/G
Traceback (most recent call last):
  File "/media/local-disk2/helgeland/swiss-test/siwss/bin/swiss", line 20, in <module>
    main()
  File "/media/local-disk2/helgeland/swiss-test/siwss/local/lib/python2.7/site-packages/swiss/main.py", line 1022, in main
    run_process(opts.assoc,opts.trait,opts.out,opts)
  File "/media/local-disk2/helgeland/swiss-test/siwss/local/lib/python2.7/site-packages/swiss/main.py", line 799, in run_process
    clump_results = clumper.ld_clump(opts.clump_ld_thresh,opts.clump_ld_dist)
  File "/media/local-disk2/helgeland/swiss-test/siwss/local/lib/python2.7/site-packages/swiss/LDClumper.py", line 71, in ld_clump
    ld_ok = self.finder.compute(top_marker,top_chr,top_pos - ld_dist,top_pos + ld_dist,ld_thresh)
  File "/media/local-disk2/helgeland/swiss-test/siwss/local/lib/python2.7/site-packages/swiss/PlinkLDFinder.py", line 222, in compute
    self.data = self._run_ld()
  File "/media/local-disk2/helgeland/swiss-test/siwss/local/lib/python2.7/site-packages/swiss/PlinkLDFinder.py", line 241, in _run_ld
    data = self._ld_refvar_region(vcf,self.variant,region,min_r2=self.min_r2,ignore_filter=True)
  File "/media/local-disk2/helgeland/swiss-test/siwss/local/lib/python2.7/site-packages/swiss/PlinkLDFinder.py", line 343, in _ld_refvar_region
    raise Exception, "\n" + ld_stdout + "\n\n" + ld_stderr
Exception:
PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/
(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /media/local-disk2/helgeland/swiss-test/tmpu3nn9N.log.
Options in effect:
  --ld-snp id0
  --ld-window 99999
  --ld-window-kb 99999
  --ld-window-r2 0.8
  --out /media/local-disk2/helgeland/swiss-test/tmpu3nn9N
  --r2 gz dprime with-freqs yes-really
  --threads 1
  --vcf /dev/fd/0

128829 MB RAM detected; reserving 64414 MB for main workspace.

Error: File read failure.
```

**Comment by welchr** [comment]
Apologies for the delay...

I added some debug output that may (or may not) shed some light on the problem. If you have time, install the development code by doing: 

```
pip install --no-deps -I --force-reinstall git+https://github.com/welchr/swiss.git@bug/plink-read-failure
```

And then re-run your command with the `--debug` flag added and post the output. Note this command will take longer than you're used to, you might have to let it run for quite a while.

**Comment by welchr** [comment]
Closing for now - will reopen if activity.

**Comment by wendyxwang** [comment]
Hi,

I wonder if this issue has been solved? I ran into the same error message where plink can not calculate LD for any SNP, but it didn't break the program, and swiss went through all SNPs with a "fail" in ld_clump.

**Comment by welchr** [comment]
If you're able, try running the install command above and attach the debug output. That branch adds some extra debugging information to help track down the problem.

**Comment by wendyxwang** [comment]
The install command above didn't provide more information to track down the
error message. I got the same errors as before.

**Comment by welchr** [comment]
Can you post the output?

**Comment by wendyxwang** [comment]
The beginning of log file looks like following:
Loaded 12740 variants (after filters) from association results..

LD clumping results..
- LD source: vcfmap.json
- LD source file: vcfmap.json
- LD threshold: 0.1 (r2)
- LD max distance: 1,000,000 (bp)

Starting with significant 12740 SNPs:
variant             P CHR        POS
                          chr1:1127664_A/G  6.846000e-06   1    1127664
                          chr1:1130902_A/G  5.574000e-06   1    1130902
                         chr1:12224083_G/T  2.226000e-06   1   12224083
                     chr1:19961780_GAATC/G  1.905000e-06   1   19961780
                         chr1:19995392_A/G  1.111000e-07   1   19995392

And for every variant, the message reads as following:

Working on LD clumping variant chr18:48925957_A/G
PLINK v1.90b1g 64-bit (10 Jun 2014)        https://www.cog-genomics.org/plink2
(C) 2005-2014 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to work directory/tmpInGH0o.log.
386593 MB RAM detected; reserving 193296 MB for main workspace.

Error: File read failure.

[33mWarning: [0mcould not calculate LD for variant chr18:48925957_A/G at 18:48925957

**Comment by wendyxwang** [comment]
We suspect part of the problem is that empty input is being fed to plink, which might be because the chromosome in the vcfs has "chr" instead of a number.

**Comment by welchr** [comment]
The previous user's VCF/results didn't have chromosomes prefixed with chr and ran into the same error. I think the code drops them anyway. 

Are you able to share a small chunk of your data for debugging? 

Also, could you try upgrading your PLINK 1.9 to the latest version to see if that helps? Looks like you're on an old one from June 2014.

**Comment by wendyxwang** [comment]
This is the only version of PLINK we have installed. But we can definitely install a newer version. Which verison of PLINK would you specifically recommend?

We can create a mock vcf file and share it.

**Comment by welchr** [comment]
Unfortunately I think you're stuck using the latest version, since they don't make previous versions available for download (insofar as I'm aware, anyway.) Try the stable build at this link: 

https://www.cog-genomics.org/plink2

**Comment by wendyxwang** [comment]
We found that there maybe two issues:
1. tabix failed because we didn't know tbl files are needed. But it would be helpful if the error message can direct users to it.
2. We saw that tabix command was:
`/usr/bin/tabix -h /fh/scratch/delete30/peters_u/krcurtis/t632-work/extracted_vcfs/topmedLD_freeze6_chr18.vcf.gz 18:47925503-49925503`
However, all variants have "chr" in their chromosome label. So this above command line produces a vcf with no variants. If we change it to this to try it on command line:
`/usr/bin/tabix -h /fh/scratch/delete30/peters_u/krcurtis/t632-work/extracted_vcfs/topmedLD_freeze6_chr18.vcf.gz chr18:47925503-49925503`
Then, tabix produce vcfs with variants.

**Comment by welchr** [comment]
Awesome, that’s really helpful. I can patch both of those in the next few days. Also happy to merge in a PR if you would like to patch it on your end, either way is good with me.

**Comment by wendyxwang** [comment]
That would be great if you can path them and I will run swiss again once it's updated, to see if this issue get fixed. Thanks for being so responsive!

**Comment by wendyxwang** [comment]
Hi Ryan,

I'm just checking if there is any progress on this issue? Hope this is the issue that cause error messages.

Thanks,
Wendy

**Comment by welchr** [comment]
Apologies, Wendy. We've had some large issues on another project that have been eating my time. Hopefully I can finish by the end of the week. I was able to patch the first problem listed above, and had started implementing the second when I was interrupted. Please ping me again if you don't hear something soon.

**Comment by wendyxwang** [comment]
No worries. Thanks for updates!

**Comment by welchr** [comment]
@wendyxwang Hopefully this takes care of the issue - can you try installing using the test branch (new command below)? 

```
pip install --no-deps -I --force-reinstall git+https://github.com/statgen/swiss.git@fix/chr-prefix
```

**Comment by welchr** [comment]
Fixed in v1.1.0: 

* de508c899be5728d0768f0c70216194f12d4e202
* aa46e1724bd2e45a33e695b0718dda4e0f6a35b0
